python/ray/ray_logging.py

```diff
@@ -58,3 +71,120 @@ def isatty(self):
         # Return the standard out isatty. This is used by colorful.
         fd = 1 if self.intercept_stdout else 2
         return os.isatty(fd)
+
+    def close(self):
+        handler = self.logger.handlers[0]
+        handler.close()
+
+    def fileno(self):
+        handler = self.logger.handlers[0]
+        return handler.stream.fileno()
+
+
+class StandardFdRedirectionRotatingFileHandler(RotatingFileHandler):
+    """RotatingFileHandler that redirects stdout and stderr to the log file.
+
+    It is specifically used to default_worker.py.
+
+    The only difference from this handler vs original RotatingFileHandler is
+    that it actually duplicates the OS level fd using os.dup2.
+    """
+
+    def __init__(self,
+                 filename,
+                 mode="a",
+                 maxBytes=0,
+                 backupCount=0,
+                 encoding=None,
+                 delay=False,
+                 is_for_stdout=True):
+        super().__init__(
+            filename,
+            mode=mode,
+            maxBytes=maxBytes,
+            backupCount=backupCount,
+            encoding=encoding,
+            delay=delay)
+        self.is_for_stdout = is_for_stdout
+        self.switch_os_fd()
+
+    def doRollover(self):
+        super().doRollover()
+        self.switch_os_fd()
+
+    def get_original_stream(self):
+        if self.is_for_stdout:
+            return sys.stdout
+        else:
+            return sys.stderr
+
+    def switch_os_fd(self):
+        # Old fd will automatically closed by dup2 when necessary.
```

Comment:
I seem to recall having to call `sys.stdout = open(..., closefd=False)` before to avoid some issues with flushing python's high level buffer. Do you think we may still need to do that?

Reply:
I guess not because sys.stdout will be always the interceptor class. I will make sure if this will work in the next PR (I will add regression tests as well as log rotation test in the next PR).